File: runtime/ruby/lib/agentic_engine/runner.rb
# frozen_string_literal: true

module AgenticEngine
  # Executes a single agent (LLM or deterministic) and returns an AgentResult.
  # Maintains the handler registry for deterministic agents.
  module Runner
    class RunnerError < StandardError; end

    @handlers = {} # { "handler_name" => callable }
    @log = Logger.create("runner")

    class << self
      attr_reader :handlers

      # Register a deterministic handler.
      #
      # @param name [String] handler name (must match agent.yaml handler field)
      # @param callable [#call] a Proc, lambda, or any object responding to #call.
      #   Receives a single Hash argument (the resolved input) and must return a
      #   Hash with :output and :metrics keys (or just :output).
      def register_handler(name, callable)
        @log.info("register_handler", { name: name })
        @handlers[name] = callable
      end

      # Reset handler registry (useful in tests).
      def reset_handlers!
        @handlers.clear
      end

      # Execute an agent given its definition and resolved input hash.
      #
      # @param input [Hash] resolved input parameters
      # @param agent_def [AgentDefinition]
      # @param config_overrides [Hash, nil] optional overrides for model/temperature
      # @return [Hash] { output:, metrics: } (plain hashes for JSON compat)
      def execute_agent(input, agent_def, config_overrides: nil)
        case agent_def.type
        when "llm"
          execute_llm_agent(input, agent_def, config_overrides: config_overrides)
        when "deterministic"
          execute_deterministic_agent(input, agent_def)
        else
          raise RunnerError, "Unknown agent type: #{agent_def.type}"
        end
      end

      # Convenience method: load an agent by ID and execute it.
      #
      # @param agent_id [String] agent directory name
      # @param input [Hash] resolved input parameters
      # @param config_overrides [Hash, nil]
      # @return [Hash] { output:, metrics: }
      def invoke_agent(agent_id, input, config_overrides: nil)
        agent_def = Loader.load_agent(agent_id)
        execute_agent(input, agent_def, config_overrides: config_overrides)
      end

      private

      def execute_llm_agent(input, agent_def, config_overrides: nil)
        model = (config_overrides && config_overrides["model"]) || agent_def.model
        temperature = (config_overrides && config_overrides["temperature"]) || agent_def.temperature || 0.7

        raise RunnerError, "LLM agent '#{agent_def.name}' has no model configured" unless model

        # Build user content from template + input
        user_content = if agent_def.user_message
                         Resolver.resolve_template(agent_def.user_message, input)
                       else
                         JSON.generate(input)
                       end

        @log.debug("execute_llm_agent", { agent: agent_def.name, model: model })

        LLM.call_llm(
          model:         model,
          system_prompt: agent_def.prompt,
          user_content:  user_content,
          temperature:   temperature,
          schema_name:   agent_def.schema
        )
      end

      def execute_deterministic_agent(input, agent_def)
        handler_name = agent_def.handler
        raise RunnerError, "Deterministic agent '#{agent_def.name}' has no handler" unless handler_name

        callable = @handlers[handler_name]
        raise RunnerError, "No handler registered for '#{handler_name}'. " \
                           "Call AgenticEngine.register_handler('#{handler_name}', your_proc) first." unless callable

        @log.debug("execute_deterministic_agent", { agent: agent_def.name, handler: handler_name })

        start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)
        result = callable.call(input)
        end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)

        # Normalize the result: handler can return a plain hash (treated as
        # output) or a hash with :output/:metrics keys.
        if result.is_a?(Hash) && (result.key?(:output) || result.key?("output"))
          output  = result[:output] || result["output"]
          metrics = result[:metrics] || result["metrics"] || {
            latency_ms: (end_time - start_time).to_i,
            input_tokens: 0,
            output_tokens: 0
          }
        else
          output  = result
          metrics = {
            latency_ms: (end_time - start_time).to_i,
            input_tokens: 0,
            output_tokens: 0
          }
        end

        { output: output, metrics: metrics }
      end
    end
  end
end

File: runtime/python/agentic_engine/runner.py
"""Agent execution — deterministic handlers and LLM agents.

Deterministic handlers are registered at module level via
``register_handler(name, fn)``.  LLM agents build a user message from the
template defined in their agent YAML then call the LLM.
"""

from __future__ import annotations

import asyncio
import inspect
from typing import Any, Awaitable, Callable

from .llm import call_llm
from .loader import load_agent
from .logger import create_logger, serialize_error
from .resolver import resolve_template
from .types import AgentDefinition, AgentResult, StepMetrics

log = create_logger("runner")

# ---------------------------------------------------------------------------
# Handler registry
# ---------------------------------------------------------------------------

HandlerFn = Callable[..., Any]
_handler_registry: dict[str, HandlerFn] = {}


def register_handler(name: str, fn: HandlerFn) -> None:
    """Register a deterministic handler function by name.

    The function may be sync or async.  It will receive a single ``input``
    dict and must return a dict.
    """
    _handler_registry[name] = fn
    log.debug("Registered handler", name=name)


# ---------------------------------------------------------------------------
# Agent execution
# ---------------------------------------------------------------------------

async def execute_agent(
    input_data: dict[str, Any],
    agent_def: AgentDefinition,
    config_overrides: dict[str, Any] | None = None,
) -> AgentResult:
    """Execute a single agent given its definition and resolved input.

    For *deterministic* agents the registered handler is called directly.
    For *llm* agents the user message template is rendered and the LLM is
    invoked.

    Parameters
    ----------
    input_data:
        Resolved input dict for the agent.
    agent_def:
        The agent definition (from loader).
    config_overrides:
        Optional dict with keys like ``model`` or ``temperature`` that
        override the agent definition values for this execution.
    """
    overrides = config_overrides or {}
    model = overrides.get("model", agent_def.model)
    temperature = float(overrides.get("temperature", agent_def.temperature))

    if agent_def.type == "deterministic":
        return await _run_deterministic(agent_def.handler, input_data)
    elif agent_def.type == "llm":
        return await _run_llm(
            agent_def=agent_def,
            input_data=input_data,
            model=model,
            temperature=temperature,
        )
    else:
        raise ValueError(f"Unknown agent type: {agent_def.type}")


async def _run_deterministic(
    handler_name: str, input_data: dict[str, Any]
) -> AgentResult:
    """Invoke a registered deterministic handler."""
    if handler_name not in _handler_registry:
        raise KeyError(
            f"No handler registered for '{handler_name}'. "
            f"Call register_handler('{handler_name}', fn) before invoking."
        )

    handler = _handler_registry[handler_name]
    log.info("Running deterministic handler", handler=handler_name)

    import time

    start = time.monotonic()
    try:
        result = handler(input_data)
        # Support async handlers transparently.
        if inspect.isawaitable(result):
            result = await result
    except Exception as exc:
        log.error("Handler failed", handler=handler_name, **serialize_error(exc))
        raise

    elapsed_ms = (time.monotonic() - start) * 1000
    metrics = StepMetrics(duration_ms=round(elapsed_ms, 2))

    return AgentResult(output=result, metrics=metrics)


async def _run_llm(
    agent_def: AgentDefinition,
    input_data: dict[str, Any],
    model: str,
    temperature: float,
) -> AgentResult:
    """Build the user message from the template and call the LLM."""
    user_message = resolve_template(agent_def.user_message, input_data)

    output, metrics = await call_llm(
        model=model,
        system_prompt=agent_def.system_prompt,
        user_content=user_message,
        temperature=temperature,
        schema_name=agent_def.schema,
        input_type=agent_def.input_type,
        image_detail=agent_def.image_detail,
    )

    return AgentResult(output=output, metrics=metrics)


# ---------------------------------------------------------------------------
# Convenience: invoke by agent ID
# ---------------------------------------------------------------------------

async def invoke_agent(
    agent_id: str,
    input_data: dict[str, Any],
    config_overrides: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Load an agent by ID, execute it, and return the result as a dict.

    This is the primary public entry point for one-shot agent invocations.
    """
    agent_def = load_agent(agent_id)
    result = await execute_agent(input_data, agent_def, config_overrides)
    return {
        "output": result.output,
        "metrics": {
            "duration_ms": result.metrics.duration_ms,
            "prompt_tokens": result.metrics.prompt_tokens,
            "completion_tokens": result.metrics.completion_tokens,
            "total_tokens": result.metrics.total_tokens,
        },
    }
